app/controllers/admin/take_part_pages_controller.rb

```diff
@@ -53,4 +55,20 @@ def take_part_page_params
       :title, :summary, :body, :image, :image_alt_text, :image_cache
     )
   end
+
+  def patch_getinvolved_page_links
```

Comment:
At the moment, this method isn't being called when a TakePartPage is created or deleted, which means the links will become out of sync unless another one is updated or reordered.

It might be an idea to move the `patch_getinvolved_page_links` method to be a callback (e.g. [`after_commit`](https://api.rubyonrails.org/v6.1.4/classes/ActiveRecord/Transactions/ClassMethods.html#method-i-after_commit)) on the `TakePartPage` model.  Then it'll also cover the create and destroy methods, and you won't need to call it at all in the controller.

Reply:
Added this as an "after_action" call in the Controller as this was the most logical place to put it that ensured everything was being updated correctly - the model didn't want to note certain changes.

Reply:
I don't think `after_action` is necessarily the right callback to use here, as that will execute after every action (including `index` and `show`, where a view is being rendered and no data being changed).

What was the issue with adding this to the model?  I think that would be a nicer solution, as it would ensure the links get updated regardless of which controller makes the changes.

Reply:
The model doesn't fire the `after_commit` upon updating orders, for one. You can actually define which actions trigger the execution with [an `on:` definition](https://api.rubyonrails.org/v6.1.4/classes/ActiveRecord/Transactions/ClassMethods.html#method-i-after_commit) which for me may be the best solution here? I meant to put it in place but I wanted to see your thoughts.

Reply:
The `after_commit` doesn't get called when the pages are redordered because there is a `update_column` [here](https://github.com/alphagov/whitehall/commit/1100ced7622cecbab9989a86a8c769bafb6cafa9#diff-4553ff371557c14602c2d83baea491d31490b06f0be595b350cf26fad5f6e392R48) which [skips callbacks](https://apidock.com/rails/v3.2.8/ActiveRecord/Persistence/update_column).

If you change that line to `page.update(ordering: ids_in_new_ordering.index(page.id.to_s) + 1)`, you should find the callback won't be skipped.

As the reordering happens in a transaction, the `after_commit` will only get called once at the end of the commit, after the items have been correctly placed in order in the database.  I tried this locally on a Rails console (adding a `puts "patching"` statement into `patch_getinvolved_page_links` method to see when this actually occurs) and you can see the patching only takes place after all the reordering has taken place:
```
[21] pry(main)> TakePartPage.reorder!([3,1,2])
  TRANSACTION (0.4ms)  BEGIN
  TakePartPage Load (0.5ms)  SELECT `take_part_pages`.* FROM `take_part_pages` WHERE `take_part_pages`.`id` IN (3, 1, 2) ORDER BY `take_part_pages`.`id` ASC LIMIT 1000
  TakePartPage Update (0.4ms)  UPDATE `take_part_pages` SET `take_part_pages`.`ordering` = 2, `take_part_pages`.`updated_at` = '2021-09-01 14:38:14' WHERE `take_part_pages`.`id` = 1
  TakePartPage Load (0.4ms)  SELECT `take_part_pages`.* FROM `take_part_pages` WHERE `take_part_pages`.`id` = 1 LIMIT 1
  TakePartPage Update (0.4ms)  UPDATE `take_part_pages` SET `take_part_pages`.`ordering` = 3, `take_part_pages`.`updated_at` = '2021-09-01 14:38:14' WHERE `take_part_pages`.`id` = 2
  TakePartPage Load (0.4ms)  SELECT `take_part_pages`.* FROM `take_part_pages` WHERE `take_part_pages`.`id` = 2 LIMIT 1
  TakePartPage Update (0.4ms)  UPDATE `take_part_pages` SET `take_part_pages`.`ordering` = 1, `take_part_pages`.`updated_at` = '2021-09-01 14:38:14' WHERE `take_part_pages`.`id` = 3
  TakePartPage Load (0.5ms)  SELECT `take_part_pages`.* FROM `take_part_pages` WHERE `take_part_pages`.`id` = 3 LIMIT 1
  TakePartPage Update All (0.4ms)  UPDATE `take_part_pages` SET `take_part_pages`.`ordering` = 4 WHERE (id NOT IN ('3','1','2'))
  TRANSACTION (1.3ms)  COMMIT
patching
  TakePartPage Load (0.5ms)  SELECT `take_part_pages`.* FROM `take_part_pages` ORDER BY `take_part_pages`.`ordering` ASC
```

Reply:
Oooh perfect, thank you, I'll throw this in.

Reply:
Right this now works as designed again - now including edits, deletions and additions, the only issue is [within the rake task](https://github.com/alphagov/whitehall/blob/main/lib/tasks/publishing_api.rake#L98) - it doesn't update the patching. That is a separate piece of work I think and can be approached later.

Reply:
In theory, re-publishing the Take Part page to the Publishing API should result in link expansion updating everywhere the document is linked to, so hopefully nothing for us to worry about for now!